src/java/org/apache/cassandra/index/sai/disk/v1/trie/TrieTermsDictionaryReader.java

```diff
@@ -0,0 +1,284 @@
+    {
+        @Override
+        public int sizeofNode(SerializationNode<Long> node, long nodePosition)
+        {
+            return TrieNode.typeFor(node, nodePosition).sizeofNode(node) + sizeof(node.payload());
+        }
+
+        @Override
+        public void write(DataOutputPlus dest, SerializationNode<Long> node, long nodePosition) throws IOException
+        {
+            final TrieNode type = TrieNode.typeFor(node, nodePosition);
+            final Long payload = node.payload();
+            if (payload != null)
+            {
+                final int payloadBits = SizedInts.nonZeroSize(payload);
+                type.serialize(dest, node, payloadBits, nodePosition);
+                SizedInts.write(dest, payload, payloadBits);
+            }
+            else
+            {
+                type.serialize(dest, node, 0, nodePosition);
+            }
+        }
+
+        private int sizeof(Long payload)
+        {
+            if (payload != null)
+            {
+                return SizedInts.nonZeroSize(payload);
+            }
+            return 0;
+        }
+    };
+
+    public long exactMatch(ByteComparable key)
+    {
+        int b = follow(key);
+        if (b != ByteSource.END_OF_STREAM)
+        {
+            return NOT_FOUND;
+        }
+        return getCurrentPayload();
+    }
+
+    public Iterator<Pair<ByteComparable, Long>> iterator()
+    {
+        return new AbstractIterator<Pair<ByteComparable, Long>>()
+        {
+            final TransitionBytesCollector collector = new TransitionBytesCollector();
+            IterationPosition stack = new IterationPosition(root, null);
+
+            @Override
+            protected Pair<ByteComparable, Long> computeNext()
+            {
+                final long node = advanceNode();
+                if (node == -1)
+                {
+                    return endOfData();
+                }
+                return Pair.create(collector.toByteComparable(), getCurrentPayload());
+            }
+
+            private long advanceNode()
+            {
+                long child;
+                int transitionByte;
+
+                go(stack.node);
+                while (true)
+                {
+                    int childIndex = stack.childIndex + 1;
+                    transitionByte = transitionByte(childIndex);
+
+                    if (transitionByte > 256)
+                    {
+                        // ascend
+                        stack = stack.prev;
+                        collector.pop();
+                        if (stack == null)
+                        {
+                            // exhausted whole trie
+                            return -1;
+                        }
+                        go(stack.node);
+                        continue;
+                    }
+
+                    child = transition(childIndex);
```

Comment:
You could take the `stack.childIndex = childIndex;` and add it here...then completely remove the `else` block below...I think :)